cleverhans/attacks.py

```diff
@@ -490,80 +490,42 @@ def __init__(self, model, back='tf', sess=None):
-
-        # Define appropriate graph (targeted / random target labels)
-        if self.y_target is not None:
-            def jsma_wrap(x_val, y_target):
-                return jsma_batch(self.sess, x, preds, grads, x_val,
-                                  self.theta, self.gamma, self.clip_min,
-                                  self.clip_max, nb_classes,
-                                  y_target=y_target)
-
-            # Attack is targeted, target placeholder will need to be fed
-            wrap = tf.py_func(jsma_wrap, [x, self.y_target], tf.float32)
-        else:
-            def jsma_wrap(x_val):
-                return jsma_batch(self.sess, x, preds, grads, x_val,
-                                  self.theta, self.gamma, self.clip_min,
-                                  self.clip_max, nb_classes,
-                                  y_target=None)
-
-            # Attack is untargeted, target values will be chosen at random
-            wrap = tf.py_func(jsma_wrap, [x], tf.float32)
-
-        return wrap
-
-    def parse_params(self, theta=1., gamma=np.inf, nb_classes=None,
-                     clip_min=0., clip_max=1., y_target=None, **kwargs):
-        """
-        Take in a dictionary of parameters and applies attack-specific checks
-        before saving them as attributes.
-
-        Attack-specific parameters:
         :param theta: (optional float) Perturbation introduced to modified
-                      components (can be positive or negative)
-        :param gamma: (optional float) Maximum percentage of perturbed features
-        :param nb_classes: (optional int) Number of model output classes
+            components (can be positive or negative)
+        :param epochs: (optional, int or float) maximum epochs to run.  If
+            this is a floating number between [0, 1], then it is treated as
+            gamma, the distortion factor in the paper, and the max epochs is
+            automatically calculated.
+        :param pair: If true, add noise to two pixels at a time.  Otherwise
+            one pixel at a time.
+        :param min_proba: The minimum desired probability that a model will
+            wrongly classify the adversarials.
         :param clip_min: (optional float) Minimum component value for clipping
         :param clip_max: (optional float) Maximum component value for clipping
-        :param y_target: (optional) Target tensor if the attack is targeted
         """
 
-        if nb_classes is not None:
-            warnings.warn("The nb_classes argument is depricated and will "
-                          "be removed on 2018-02-11")
-        self.theta = theta
-        self.gamma = gamma
-        self.clip_min = clip_min
-        self.clip_max = clip_max
-        self.y_target = y_target
+        y = kwargs.get('y_target', 0)
+        epochs = kwargs.get('epochs', 0.2)
+        theta = kwargs.get('theta', 1.0)
+        pair = kwargs.get('pair', True)
```

Comment:
See comment about generalizing to `k` in discussion. For this PR, it's OK to have the code throw an error when `k>2`

Reply:
I will do that.  And I'm not considering implementing a general case.  Since it is unlikely that JSMA will be anything practical for `k>2` case.  For an image with `28x28`, each iteration goes through ~1M pixels for `k=2`, ~1B pixels for `k=3`.